extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/HttpUpgradePolicy.java

```diff
@@ -0,0 +1,98 @@
+package io.quarkus.websockets.next;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import io.quarkus.security.identity.SecurityIdentity;
+import io.smallrye.mutiny.Uni;
+import io.vertx.core.http.HttpServerRequest;
+
+/**
+ * A policy that controls which requests are allowed to upgrade the HTTP connection to a WebSocket connection.
+ * CDI beans implementing this interface are invoked on every request.
+ * The CDI beans implementing `HttpUpgradePolicy` interface can be either `@ApplicationScoped`, `@Singleton`
+ * or `@Dependent` beans, but never the `@RequestScoped` beans.
+ * <p>
+ * The policies are called orderly according to a bean priority.
+ * When no priority is declared (for example with the `@jakarta.annotation.Priority` annotation), default priority is used.
+ * If one of the policies rejects the upgrade, remaining policies are not called.
+ */
+public interface HttpUpgradePolicy {
+
+    /**
+     * This method inspects HTTP Upgrade context and either allows or denies upgrade to a WebSocket connection.
+     *
+     * @param context {@link HttpUpgradeContext}
+     * @return check result; must never be null
+     */
+    Uni<CheckResult> checkUpgrade(HttpUpgradeContext context);
+
+    /**
+     * Determines WebSocket endpoints this policy is applied to.
+     *
+     * @param endpointId WebSocket endpoint id, @see {@link WebSocket#endpointId()} for more information
+     * @return true if this policy should be applied on a WebSocket endpoint with given id
+     */
+    default boolean appliesTo(String endpointId) {
+        return true;
+    }
+
+    /**
+     * @param httpRequest {@link HttpServerRequest}; the HTTP 1.X request employing the 'Upgrade' header
+     * @param securityIdentity {@link SecurityIdentity}; the identity is null if the Quarkus Security extension is absent
+     */
+    record HttpUpgradeContext(HttpServerRequest httpRequest, SecurityIdentity securityIdentity) {
+    }
+
+    final class CheckResult {
+
+        private static final CheckResult PERMIT_UPGRADE = new CheckResult(true, null, null);
+
+        private final boolean upgradePermitted;
+        private final int httpResponseCode;
+        private final Map<String, String> responseHeaders;
```

Comment:
This should be actually `Map<String,List<String>>` because you can set multiple header values for a specific header name.

Reply:
I know, didn't want to complicate it, but sure, I'll implement it.

Reply:
done, we could add some utility methods for simple scenarios (in the future); I also tried to avoid Vert.x MultiMap even though this extension depends on it, I am not sure it is really part of the API.

Reply:
I agree that we should not really use the `MultiMap` from Vert.x.